plugins/grounded/grounded.cpp

```diff
@@ -0,0 +1,177 @@
+
+	return wrapper;
+}
+
+MumbleStringWrapper mumble_getDescription() {
+	static const char description[] = "Positional audio support for Grounded. Steam release version 1.4.3.4578.";
+
+	MumbleStringWrapper wrapper;
+	wrapper.data           = description;
+	wrapper.size           = strlen(description);
+	wrapper.needsReleasing = false;
+
+	return wrapper;
+}
+
+MumbleStringWrapper mumble_getAuthor() {
+	static const char author[] = "MumbleDevelopers";
+
+	MumbleStringWrapper wrapper;
+	wrapper.data           = author;
+	wrapper.size           = strlen(author);
+	wrapper.needsReleasing = false;
+
+	return wrapper;
+}
+
+mumble_version_t mumble_getAPIVersion() {
+	return MUMBLE_PLUGIN_API_VERSION;
+}
+
+void mumble_registerAPIFunctions(void *) {
+}
+
+void mumble_releaseResource(const void *) {
+}
+
+mumble_version_t mumble_getVersion() {
+	return { 1, 0, 0 };
+}
+
+uint32_t mumble_getFeatures() {
+	return MUMBLE_FEATURE_POSITIONAL;
+}
+
+uint8_t mumble_initPositionalData(const char *const *programNames, const uint64_t *programPIDs, size_t programCount) {
+	for (size_t i = 0; i < programCount; ++i) {
+		if (strcmp(programNames[i], "Maine-Win64-Shipping.exe") != 0) {
+			continue;
+		}
+
+		ProcessWindows proc(programPIDs[i], programNames[i]);
+
+		if (!proc.isOk()) {
+			continue;
+		}
+
+		const Modules &modules = proc.modules();
+		const auto iter        = modules.find("Maine-Win64-Shipping.exe");
+
+		if (iter == modules.cend()) {
+			continue;
+		}
+
+		procptr_t p = iter->second.baseAddress();
+
+		if (!(p = proc.peekPtr(p + 0x0612cb38))) {
+			continue;
+		}
+
+		if (!(p = proc.peekPtr(p + 0x0))) {
+			continue;
+		}
+
+		if (!(p = proc.peekPtr(p + 0x8))) {
+			continue;
+		}
+
+		p += 0x8c0;
+
+		GroundedCam _cam;
+
+		if (!(proc.peek(p, _cam))) {
+			continue;
+		}
+
+		handle = std::make_unique< GroundedHandle >(GroundedHandle{ std::move(proc), p });
+
+		return MUMBLE_PDEC_OK;
+	}
+
+	return MUMBLE_PDEC_ERROR_TEMP;
+}
+
+void mumble_shutdownPositionalData() {
+	handle.reset();
+}
+
+bool mumble_fetchPositionalData(float *avatarPos, float *avatarDir, float *avatarAxis, float *cameraPos,
+								float *cameraDir, float *cameraAxis, const char **contextPtr,
+								const char **identityPtr) {
+	*contextPtr  = "";
+	*identityPtr = "";
+
+	ProcessWindows &proc = std::get< 0 >(*handle);
+	procptr_t camAddr    = std::get< 1 >(*handle);
+
+	GroundedCam cam;
+
+	if (!proc.peek< GroundedCam >(camAddr, cam)) {
+		std::fill_n(avatarPos, 3, 0.f);
+		std::fill_n(avatarDir, 3, 0.f);
+		std::fill_n(avatarAxis, 3, 0.f);
+
+		std::fill_n(cameraPos, 3, 0.f);
+		std::fill_n(cameraDir, 3, 0.f);
+		std::fill_n(cameraAxis, 3, 0.f);
+
+		return false;
+	}
+
+	avatarAxis[0] = cameraAxis[0] = -cam.top[0];
+	avatarAxis[1] = cameraAxis[1] = cam.top[2];
+	avatarAxis[2] = cameraAxis[2] = -cam.top[1];
+
+	avatarDir[0] = cameraDir[0] = -cam.front[0];
+	avatarDir[1] = cameraDir[1] = cam.front[2];
+	avatarDir[2] = cameraDir[2] = -cam.front[1];

```

Comment:
why don't we need a unit conversion for these fields?

Reply:
Trigonometry. The range of values goes from -1 to 1.

Let's take `cam.front[2]` for example: if you look at the sky it will be ~`1`, if you look at the ground it will be ~`-1`.

Reply:
They're unit vectors. They have a length of one in both the game and in mumble.

Reply:
I see. In that case adding an explicit `assert` that checks their magnitude to not differ significantly from one would make the code more self-explanatory.
Also, a short comment could be added that these are unitless (direction only)

Reply:
Yeah I can add comments that these should be unit vectors.

I couldn't find any other plugins to reference that do runtime assertions for these values. My understanding is a failed assertion typically causes the program to exit. If that happens, because the memory we read from the game isn't what we expect, we probably don't want mumble to quit. So I'm not sure an assert is the best choice here?

Instead, I could add a sanity check to initPositionalData that checks that the top and front vectors are of length zero (at the main menu) or one (in a game)?

Alternatively, in the general case, mumble could check that the vectors from fetch functions in positional audio plugins are of valid lengths and possibly respond by disabling the plugin if they appear to be misbehaving.

Reply:
> I couldn't find any other plugins to reference that do runtime assertions for these values. My understanding is a failed assertion typically causes the program to exit. If that happens, because the memory we read from the game isn't what we expect, we probably don't want mumble to quit. So I'm not sure an assert is the best choice here?

That's exactly what happens and that's also why it's a good thing (the crash happens right at the place where an assumption is violated - including information about where exactly the error occurred). Note however, that asserts are only active in debug builds.

Reply:
I put in an assert.

But I think is worth mentioning that, depending on how memory is written and read, it seems maybe possible to me that mumble could read a copy of memory while these vectors are being updated by the game and partially written for a particular frame. My guess is that each dimension in the three-dimensional vectors are updated as separate instructions. And, though the timing would need to be _very_ unfortunate, it's possible mumble could read a copy of this memory in the middle of an update. So one dimension might be accurate for the current frame and the others dimensions for the previous frame. In that case, the vector that mumble reads might not be a unit vector since it isn't accurate to one particular game state. Here, there isn't really anything mumble can do except maybe discard that fetch and/or try to read the memory again. I don't know that exiting is super useful, but ultimately that's not my decision.

On the other hand, it could make it very obvious if the plugin is attempting to be used but is out of date. So that could be nice.

Reply:
Since it will only be an issue in debug builds that will be used by developers, I wouldn't consider this to be an issue. It's true that the assertion could produce a false-positive under very specific circumstances but I think it is still better to have the assumptions encoded into the code explicitly.